### PR TITLE
Remove subscriptions

### PR DESCRIPTION
**Pull request by willjgriff**
Fixes issue #1 .

New behaviour removes the need for DAO's/their associated Agreements contracts to subscribe to the court in order to use it. It also removes the governor from the subscriptions contract, which previously collected some percent of the collected fees.

New behaviour allows anyone to donate HNY, or whatever the fee token is, to the CourtSubscriptions contract directly. The contracts fee token balance will be stored the first time someone claims within a period. That amount of fee token is then distributed amongst the jurors of the previous period using a random term during that period to determine distribution based on their active stake at that term (this prevents someone activating all stake at the end term of a period and claiming all the fees). When the next period starts, the total balance of the contract, including any unclaimed fees are marked for distribution for the period before.

Includes slightly re-written tests.

### CODE REVIEW

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 1]
```diff
@@ -52,251 +39,95 @@ contract CourtSubscriptions is ControlledRecoverable, TimeHelpers, ISubscription
...
-        _setPrePaymentPeriods(_prePaymentPeriods);
-    }
-
-    /**
-    * @notice Set new late payment penalty `_latePaymentPenaltyPct`‱ (‱ - 1/10,000)
-    * @param _latePaymentPenaltyPct New permyriad of subscription fees that will be applied as penalty for not paying during proper period
+    * @dev Tell the share fees corresponding to a juror
+    * @param _juror Address of the juror querying the owed shared fees of
+    * @return feeToken Address of the token used for the subscription fees
+    * @return jurorShare Amount of share fees owed to the given juror for the requested period
```
Comment (fabriziovigevani) [inline contracts/subscriptions/CourtSubscriptions.sol]:
```suggestion
    * @return jurorShare Amount of share fees owed to the given juror for the previous period
```

File: contracts/disputes/DisputeManager.sol [hunk 2]
```diff
@@ -13,7 +13,6 @@ import "../voting/ICRVotingOwner.sol";
 import "../treasury/ITreasury.sol";
 import "../arbitration/IArbitrable.sol";
 import "../registry/IJurorsRegistry.sol";
-import "../subscriptions/ISubscriptions.sol";
```
Comment (fabriziovigevani) [inline contracts/disputes/DisputeManager.sol:16]: I think we should also remove in this file.

```
    string private constant ERROR_SUBSCRIPTION_NOT_PAID = "DM_SUBSCRIPTION_NOT_PAID";
```

Might have forgotten to remove it in the original court 😄

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 3]
```diff
@@ -52,251 +39,95 @@ contract CourtSubscriptions is ControlledRecoverable, TimeHelpers, ISubscription
...
-        _payFees(subscriber, msg.sender, msg.sender, _periods);
-
-        subscriber.paused = false;
-        subscriber.previousDelayedPeriods = 0;
     }
 
     /**
-    * @notice Donate fees to the Court
-    * @param _amount Amount of fee tokens to be donated
+    * @notice Set new subscriptions fee to `@tokenAmount(_feeToken, _feeAmount)`
```
Comment (fabriziovigevani) [inline contracts/subscriptions/CourtSubscriptions.sol]:
```suggestion
    * @notice Set new subscriptions fee token to `_feeToken`
```

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 4]
```diff
@@ -52,251 +39,95 @@ contract CourtSubscriptions is ControlledRecoverable, TimeHelpers, ISubscription
...
+    * @dev Accumulated fees owed to governor (if any) will be transferred
```
Comment (fabriziovigevani) [inline contracts/subscriptions/CourtSubscriptions.sol]:
```suggestion
```

File: test/subscriptions/subscriptions-donations.js [hunk 5]
```diff
@@ -1,91 +1,48 @@
...
 const { buildHelper } = require('../helpers/wrappers/court')(web3, artifacts)
-const { assertRevert } = require('../helpers/asserts/assertThrow')
-const { assertAmountOfEvents } = require('../helpers/asserts/assertEvent')
-const { SUBSCRIPTIONS_ERRORS } = require('../helpers/utils/errors')
-const { SUBSCRIPTIONS_EVENTS } = require('../helpers/utils/events')
 
 const CourtSubscriptions = artifacts.require('CourtSubscriptions')
 const ERC20 = artifacts.require('ERC20Mock')
 
+// TODO: Delete these tests?
```
Comment (fabriziovigevani) [inline test/subscriptions/subscriptions-donations.js]: I think we could delete them yes, Looks like we are just testing ERC20 token transfer now.

File: test/subscriptions/subscriptions-setters.js [hunk 6]
```diff
@@ -84,310 +36,39 @@ contract('CourtSubscriptions', ([_, governor, someone, something, subscriber]) =
         })
 
         context('when the given fee amount is greater than zero', async () => {
```
Comment (fabriziovigevani) [inline test/subscriptions/subscriptions-setters.js]: Lets update this

File: test/subscriptions/subscriptions-jurors.js [hunk 7]
```diff
@@ -13,257 +13,335 @@ const JurorsRegistry = artifacts.require('JurorsRegistry')
...
 
-              it('transfers share fees to the juror', async () => {
-                const previousBalance = await feeToken.balanceOf(juror)
+        const currentBalance = await feeToken.balanceOf(jurorPeriod0Term1)
+        assertBn(previousBalance.add(secondJurorExpectedShareFees), currentBalance, 'juror balance does not match')
+        assert.isTrue(await subscriptions.hasJurorClaimed(jurorPeriod0Term1))
+      })
 
-                await subscriptions.claimFees(periodId, { from: juror })
+      it('sets the periods details correclty', async () => {
```
Comment (fabriziovigevani) [inline test/subscriptions/subscriptions-jurors.js]:
```suggestion
      it('sets the periods details correctly', async () => {
```

**Review by fabriziovigevani — COMMENTED** [review]
Nice work man! Much simpler 😄 

From a logic perspective looks good. 
Just wondering, 
> When the next period starts, the total balance of the contract, including any unclaimed fees are marked for distribution for the period before.

Would this mean that if a juror fails to claim for a period, that unclaimed amount won't accrue for the juror and instead will be available for distribution for all jurors in the next period?

Also we should remove the `Subscriber` struct.

**Comment by willjgriff** [comment]
Thanks! :D

> Would this mean that if a juror fails to claim for a period, that unclaimed amount won't accrue for the juror and instead will be available for distribution for all jurors in the next period?

Yes, this is a limitation of the fact that conviction voting can't currently call an external function. However, since we will soon enable it too we could discuss an alternative approach at the cost of more dev time.

> Also we should remove the `Subscriber` struct.

Good spot, will remove.

Excellent reviewing once again, thank you man!

File: contracts/disputes/DisputeManager.sol (hunk 2 above)
Comment (willjgriff) [reply contracts/disputes/DisputeManager.sol:16]: I imagine I've removed in the next PR, but removed anyway.

**Review by fabriziovigevani — APPROVED** [review]
🙌 🌞

**Pull request merged**